CAT-1717 Integrate Scenes

loibi93 opened the pull request:
Introduces new Level between Projects and Sprites
Features:
- Bacpkpack
- Copy Delete Rename DragNDrop
- Automatic conversion from old to new Project structure
- Start Scene and Continue Scene Bricks
- Global Variables work in all Scenes of a Project
- Merging of two Scenes to one scene possible
- Merging of Projects changed

I ran all tests in the package locally on two emulators and two devices simultaneously for both this branch and the develop branch. I then compared the results, and rerun the tests that failed on this branch but not on the develop branch. The tests that still failed I fixed, so all of them were green.
The results can be found here: 
[Scene.Tests.zip](https://github.com/Catrobat/Catroid/files/419835/Scene.Tests.zip)

There was a GSoC meeting, where we looked over the code and reviewed it together with @robertriedl @sjaindl @robertpainsi @bernadettespieler 
Also there are some comments on the closed PR: https://github.com/Catrobat/Catroid/pull/1800
And here is some further Feedback from @robertpainsi: https://github.com/robertpainsi/Catroid/commit/4b35742537e3d179f9afa44c9b073a4d588769fb

catrobatjenkins commented:
Can a catrobat member please verify this patch? If it doesn't look evil comment with "ok to test" to start the testrun.

loibi93 commented:
Implemented [CAT-1579](https://jira.catrob.at/browse/CAT-1570) because I already had to do the necessary xml changes, so I also implemented that (was just a little bit of code).

robertpainsi commented:
:beetle:#1
Delete all projects
Go to MainMenuActivity
Press continue
Press back

No buttons visible in the MainMenuActivity

robertpainsi commented:
:beetle:#2
Go to ProjectActivity with multiple scenes
Select delete

The checkboxes should be on the left side of the scene image.

robertpainsi commented:
:beetle:#3
Create a program with two scenes (Best, use the default program and add an empty scene)
Copy both scenes
Try to scroll

Caught exception. I guess it's irrelevant, ignore.
```
08-21 01:34:54.331 11431-12709/org.catrobat.catroid E/Sprite: java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
                                                                  at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
                                                                  at java.util.ArrayList.get(ArrayList.java:308)
                                                                  at org.catrobat.catroid.content.Sprite.cloneLooks(Sprite.java:382)
                                                                  at org.catrobat.catroid.content.Sprite.clone(Sprite.java:297)
                                                                  at org.catrobat.catroid.content.Sprite.cloneForScene(Sprite.java:315)
                                                                  at org.catrobat.catroid.content.Scene.clone(Scene.java:129)
                                                                  at org.catrobat.catroid.ui.fragment.ScenesListFragment.copyScene(ScenesListFragment.java:320)
                                                                  at org.catrobat.catroid.ui.fragment.ScenesListFragment$9.run(ScenesListFragment.java:739)
                                                                  at java.lang.Thread.run(Thread.java:818)
```

Uncaught exception:
```
08-21 01:34:58.442 11431-11431/org.catrobat.catroid E/InputEventReceiver: Exception dispatching input event.
08-21 01:34:58.442 11431-11431/org.catrobat.catroid E/MessageQueue-JNI: Exception in MessageQueue callback: handleReceiveCallback
08-21 01:34:58.445 11431-11431/org.catrobat.catroid E/MessageQueue-JNI: java.lang.ClassCastException: android.widget.HeaderViewListAdapter cannot be cast to android.widget.ArrayAdapter
                                                                            at org.catrobat.catroid.ui.DynamicListView.getViewForID(DynamicListView.java:186)
                                                                            at org.catrobat.catroid.ui.DynamicListView.touchEventsEnded(DynamicListView.java:341)
                                                                            at org.catrobat.catroid.ui.DynamicListView.onTouchEvent(DynamicListView.java:250)
                                                                            ...

thmq commented:
ok to test

robertpainsi commented on catroid/src/main/java/org/catrobat/catroid/ui/dialogs/RenameSceneDialog.java:
```diff
@@ -70,7 +70,7 @@ protected boolean handleOkButton() {
 		}
 
-		if (newSceneName != null && !newSceneName.equalsIgnoreCase("")) {
+		if (!newSceneName.isEmpty() && !newSceneName.equalsIgnoreCase("")) {
```
`(!newSceneName.isEmpty() && !newSceneName.equalsIgnoreCase(""))`
 ===
`(!newSceneName.isEmpty())`

robertpainsi commented on catroid/src/main/java/org/catrobat/catroid/content/XmlHeader.java:
```diff
@@ -48,7 +49,7 @@
 	private boolean landscapeMode;
 	@SuppressWarnings("unused")
-	public boolean scenesEnabled = true;
+	public String scenesEnabled = Constants.SCENES_ENABLED_TAG;
```
Keep the boolean :wink: 

I did not test it, but I guess this member will be converted to
`<scenesEnabled>&lt;scenesEnabled&gt;</scenesEnabled>` or whatever the greater, less than signs are encoded to.
So, it's fine to use true/false.

However, previously you just used `scenesEnabled` to search for the tag which may be a string as well. When looking for a tag, use the tag symbols as well `<scenesEnabled>`.

robertpainsi commented:
Code review for the non-trivial changes can be found at https://github.com/robertpainsi/Catroid/commit/4b35742537e3d179f9afa44c9b073a4d588769fb
